### PR TITLE
Enable more warnings in purescript.cabal

### PR DESCRIPTION
**Pull request by hdgarrood**
This enables -Wincomplete-uni-patterns and -Wincomplete-record-updates
by default. They are both good warnings to have on, as they help
identify situations where code is likely to result in an error at
runtime.

-Wincomplete-uni-patterns causes a warning to be emitted for things like
this:

    head = \(x:_) -> x

which would blow up at runtime if given an empty list.

-Wincomplete-record-updates causes a warning to be emitted for code like
this:

    data Foo = Foo { x :: Int }
             | Bar

    f :: Foo -> Foo
    f foo = foo { x = 6 }

which will blow up if you try to call `f Bar`, since only the `Foo`
constructor has the `x` field. (This one isn't triggered by any of our
code at the moment, thankfully).

The reason I'm sending this now is that as of version 9.2.1, GHC
includes these two warnings as part of -Wall by default, which means
that in order to upgrade, we'd need to either rework our code to stop
emitting them, or explicitly turn them off. The former seems more
appealing to me.

Another nice consequence of this is that some functions become more
efficient, as we remove unnecessary extra traversals of lists in a
couple of cases.

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [ ] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [ ] Linked any existing issues or proposals that this pull request should close
- [ ] Updated or added relevant documentation
- [ ] Added a test for the contribution (if applicable)

### CODE REVIEW

File: src/Language/PureScript/TypeChecker/Entailment.hs
Comment (hdgarrood) [inline src/Language/PureScript/TypeChecker/Entailment.hs:874]: I opened https://github.com/haskell/core-libraries-committee/issues/67 to see if we can get this into `base`.

File: src/Language/PureScript/Sugar/Names/Env.hs
Comment (hdgarrood) [inline src/Language/PureScript/Sugar/Names/Env.hs:496]: This probably won't ever happen, but if it does, we'll get a much more useful error.

File: src/Language/PureScript/Ide/Imports.hs
Comment (hdgarrood) [inline src/Language/PureScript/Ide/Imports.hs:264]: This has the exact same behaviour as before. Using `break` here removes an unnecessary list traversal.

File: src/Language/PureScript/CoreImp/Optimizer/Unused.hs
Comment (hdgarrood) [inline src/Language/PureScript/CoreImp/Optimizer/Unused.hs:24]: This has the exact same behaviour as before; `not (any isReturn xs)` is just as expensive to compute as `break isReturn xs`, so we might as well do the latter since that also gives us access to the information we need here in case there is a return statement.

**Review by MonoidMusician — APPROVED** [review]

**Review by rhendric — APPROVED** [review]

**Comment by hdgarrood** [comment]
Just in case you aren't aware, I removed myself from the core team on GitHub so I'm no longer able to merge things by myself, even with approvals - please merge this for me when you're ready.

**Pull request merged**
